ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java

```diff
@@ -0,0 +1,895 @@
+            continue;
+          }
+        }
+      }
+      // search for duplicates in tgtIndex if required
+      if (!checkAlreadyCopied || !alreadyExistsInTgt(copyCandidate.getKey(), copyCandidateValue)) {
+        logger.trace("Adding index entry for {} because it is valid and does not already exist",
+            copyCandidate.getKey());
+        indexEntriesToCopy.add(copyCandidate);
+      }
+    }
+    // order by offset in log.
+    Collections.sort(indexEntriesToCopy, PersistentIndex.INDEX_ENTRIES_OFFSET_COMPARATOR);
+    return indexEntriesToCopy;
+  }
+
+  /**
+   * Gets all the valid index entries in the given list of index entries.
+   * @param indexSegmentStartOffset the start {@link Offset} of the {@link IndexSegment} that {@code allIndexEntries} is
+   *                                from.
+   * @param allIndexEntries the list of {@link IndexEntry} instances from which the valid entries have to be chosen.
+   * @param startOffsetOfLastIndexSegmentForDelete the start {@link Offset} of the index segment up until which delete
+   *                                                records are considered applicable.
+   * @return the list of valid entries picked from {@code allIndexEntries}.
+   * @throws StoreException if {@link BlobReadOptions} could not be obtained from the store for deleted blobs.
+   */
+  private List<IndexEntry> getValidIndexEntries(Offset indexSegmentStartOffset, List<IndexEntry> allIndexEntries,
+      Offset startOffsetOfLastIndexSegmentForDelete) throws StoreException {
+    // TODO: move this blob store stats
+    boolean deletesInEffect = startOffsetOfLastIndexSegmentForDelete != null
+        && indexSegmentStartOffset.compareTo(startOffsetOfLastIndexSegmentForDelete) <= 0;
+    List<IndexEntry> validEntries = new ArrayList<>();
+    for (IndexEntry indexEntry : allIndexEntries) {
+      IndexValue value = indexEntry.getValue();
+      if (value.isFlagSet(IndexValue.Flags.Delete_Index)) {
+        // DELETE entry. Always valid.
+        validEntries.add(indexEntry);
+        // if this delete cannot be counted and there is a corresponding unexpired PUT entry in the same index segment,
+        // we will need to add it.
+        if (!deletesInEffect && !srcIndex.isExpired(value)) {
+          long putRecordOffset = value.getOriginalMessageOffset();
+          if (putRecordOffset != IndexValue.UNKNOWN_ORIGINAL_MESSAGE_OFFSET
+              && indexSegmentStartOffset.getOffset() <= putRecordOffset) {
+            BlobReadOptions options =
+                srcIndex.getBlobReadInfo(indexEntry.getKey(), EnumSet.allOf(StoreGetOptions.class));
+            Offset offset = new Offset(indexSegmentStartOffset.getName(), putRecordOffset);
+            IndexValue putValue = new IndexValue(options.getSize(), offset, value.getExpiresAtMs());
+            validEntries.add(new IndexEntry(indexEntry.getKey(), putValue));
```

Comment:
Checking the delete entry at the start block is enough (and better to do because since it avoids a lookup if the blob is expired). Discussion regarding that [here](https://github.com/linkedin/ambry/pull/568#discussion_r109059243)